### PR TITLE
[Frontend] Kill the server on engine death

### PR DESCRIPTION
**Pull request by joerunde**
Slight mitigation for #5901 

This PR updates the open ai server to check for cases where the AsyncLLMEngine errors, and terminate itself when it does.

As I understand it, the goal of stopping the AsyncLLMEngine on unknown errors and not restarting is to prevent error loops where we try to restart the engine after an error that's really unrecoverable. However when we run into these issues in production, the server has to sit there responding with errors to all requests until a readiness probe fails, and then wait until the liveness probes fail enough times to trigger a container restart. This results longer downtime than necessary when we can instead terminate the server as soon as we know that the engine has failed.

This is implemented as a sigterm sent to the current PID which allows the uvicorn server to handle the signal and enter the usual graceful shutdown path.

There are two exception handlers added to the server. The first one checks for anything that extends `RuntimeError`, which I'm assuming will match most exceptions that kill the engine. This handler checks to see if the engine is dead, and kills the server if it is.
The second handler matches the `AsyncEngineDeadError` explicitly and immediately kills the server, to catch any cases that were not caught by the first handler. This could happen if the engine was shared with another module that accidentally killed it (not speaking from experience here or anything :wink:)

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/engine/async_llm_engine.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def _log_task_completion(task: asyncio.Task,
         error_callback(exception)
         raise AsyncEngineDeadError(
             "Task finished unexpectedly. This should never happen! "
-            "Please open an issue on Github. See stack trace above for the"
+            "Please open an issue on Github. See stack trace above for the "
```
Comment (joerunde) [inline vllm/engine/async_llm_engine.py:59]: (This has just triggered me every time I see the log)

File: vllm/entrypoints/openai/api_server.py [hunk 2]
```diff
@@ -184,6 +186,28 @@ async def validation_exception_handler(_, exc):
...
+    @app.exception_handler(AsyncEngineDeadError)
+    async def engine_dead_handler(_, __):
+        """Kill the server if the async engine is already dead. It will
+        not handle any further requests."""
+        logger.fatal("AsyncLLMEngine is already dead, terminating server "
+                     "process")
+        os.kill(os.getpid(), signal.SIGTERM)
+
+        return Response(status_code=HTTPStatus.INTERNAL_SERVER_ERROR)
+
```
Comment (Yard1) [inline vllm/entrypoints/openai/api_server.py:210]: I would recommend gating this behavior behind an env var/argument, since automatic process death may be unexpected.

Why not just raise a normal exception, too? calling `os.kill` on self seems like an anti-pattern.

Comment (joerunde) [reply vllm/entrypoints/openai/api_server.py:210]: +1 I can gate that for sure. But I do think it should be the default behavior for running the server, I don't really see the utility in leaving it up and unresponsive if we know the engine is dead.

> Why not just raise a normal exception, too? 

In this case raising an exception isn't gonna kill the server, an unhandled exception in an endpoint handler will just cause the server to respond to the request with an empty 500 (which is also what I'm explicitly doing here). The main process is waiting on the server to exit, so the server has to be notified to exit somehow. AFAIK the canonical way of doing this when using `uvicorn.run(...)` like we are is to use a termination signal. I agree that looks funky here though, It might be possible instead to do something like
```
server = uvicorn.Server(...)
server.run()
```
in our `run_server` and then use `server.shutdown()` in these handlers to try to initiate a shutdown that will then cause the main process to return from `server.run()` and exit. WDYT?

Comment (Yard1) [reply vllm/entrypoints/openai/api_server.py:210]: I think that sounds cleaner, but I don't have much experience with uvicorn, so take that with a grain of salt :)

**Review by Yard1 — APPROVED** [review]

File: vllm/entrypoints/openai/api_server.py (hunk 2 above)
Comment (joerunde) [reply vllm/entrypoints/openai/api_server.py]: Okay, I think I'm happy with the server shutdown code now, and I added a flag to gate this logic.

I'm noodling on some tests and going down the route of running `run_server()` in a thread with the llm engine patched to raise an error on each request, but it is looking a bit funky. Any other suggestions?

Comment (Yard1) [reply vllm/entrypoints/openai/api_server.py]: I think it's ok for tests

File: vllm/entrypoints/openai/api_server.py [hunk 3]
```diff
@@ -296,15 +327,24 @@ def run_server(args, llm_engine=None):
...
+                                    ssl_keyfile=args.ssl_keyfile,
+                                    ssl_certfile=args.ssl_certfile,
+                                    ssl_ca_certs=args.ssl_ca_certs,
+                                    ssl_cert_reqs=args.ssl_cert_reqs)
+
+    global server
+    server = uvicorn.Server(config=uvicorn_config)
+
+    # Run the server and block until it exits
+    with contextlib.suppress(KeyboardInterrupt):
```
Comment (dtrifiro) [inline vllm/entrypoints/openai/api_server.py:346]: Why is this needed?

File: vllm/entrypoints/openai/api_server.py (hunk 2 above)
Comment (dtrifiro) [reply vllm/entrypoints/openai/api_server.py]: Another approach that could work would be to schedule a period task that checks the engine health and stops the http server task when the engine is unhealthy. See https://github.com/vllm-project/vllm/pull/6740

Comment (joerunde) [reply vllm/entrypoints/openai/api_server.py]: @dtrifiro true, though I think in this case we can avoid a lot of unnecessary checking since we catch the errors as they happen

**Comment by joerunde** [comment]
@Yard1 I banged my head against a wall for a few hours trying to unit test this without massively refactoring the server or injecting a poison pill into the engine to trigger an error, and I'm ready to give up 😦. Do you mind merging as-is without tests?

**Comment by Yard1** [comment]
@joerunde can you outline what you tried to do?

**Comment by vrdn-23** [comment]
@joerunde @Yard1 Is this PR good to merge? We've been experiencing some issues lately where the engine dies but the server is still up and it hence doesn't restart the Kubernetes pod on its own. I'm hoping this fix will help us mitigate this issue!

**Comment by joerunde** [comment]
@Yard1 @vrdn-23 Sorry for the delay, been focusing on #6883 instead.

I went ahead and just wrote a simple test that uses a bad lora adapter to crash the server. I tried both:
- Subclassing `RemoteOpenAIServer` to use `fork` and then `mock.patch`ing out the real model loads and injecting an exception into the LLMEngine, but just ran into too many small issues to debug
- Running the equivalent of `vllm serve` in a separate thread to be able to mock out the llm engine in-process, but that turned into 150+ lines of code to maintain for this one little test which I also did not want to foist upon the rest of the maintainers

File: vllm/entrypoints/openai/cli_args.py [hunk 4]
```diff
@@ -134,6 +134,11 @@ def make_arg_parser(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
         help="When --max-logprobs is specified, represents single tokens as"
         "strings of the form 'token_id:{token_id}' so that tokens that"
         "are not JSON-encodable can be identified.")
+    parser.add_argument("--keep-alive-on-engine-death",
```
Comment (njhill) [inline vllm/entrypoints/openai/cli_args.py:137]: Only minor but I wonder if this would be better as an env var since I think it would only be used in debugging scenarios...

**Review by njhill — APPROVED** [review]
Thanks @joerunde!

File: vllm/entrypoints/openai/cli_args.py (hunk 4 above)
Comment (joerunde) [reply vllm/entrypoints/openai/cli_args.py:137]: Ah, yeah I can change that. I only recently learned that there's an explicit split between the cli args and `vllm.envs`

**Comment by joerunde** [comment]
@Yard1 It looks like the entrypoints tests pass, but other tests are failing with timeouts and it looks like I'd need to rebase main in order for the tracing tests to pass.

Do you want me to rebase and re-trigger the CI? I don't want to use up too many CI runs for this one little PR

**Comment by Yard1** [comment]
@joerunde yeah let's fix the conflicts

**Comment by joerunde** [comment]
Cool, yeah with the addition of the common `launcher` I went ahead and moved these exception handlers there so that they can have access to the server to interrupt it, and I also had to fill out the rest of the interface for the new `AsyncEngineRPCClient`

**Comment by joerunde** [comment]
Added a little bugfix for an edge case I ran into after https://github.com/vllm-project/vllm/pull/7111/files#diff-4bd825485b5162cf8021da41a2ebd3d4026929e617d1137f69bbbe4bda0ee643R208

**Comment by joerunde** [comment]
@Yard1 @njhill I think we're up to date with ci passing, mind merging?

File: vllm/entrypoints/openai/rpc/server.py [hunk 5]
```diff
@@ -96,14 +96,17 @@ async def is_server_ready(self, identity):
...
-        # Send confirmation to the client.
-        await self.socket.send_multipart([
-            identity,
-            cloudpickle.dumps(VLLM_RPC_SUCCESS_STR),
-        ])
+        try:
+            # Abort the request in the llm engine.
+            await self.engine.abort(request.request_id)
+        except Exception:
+            logger.warning("Failed to abort request %s", request.request_id)
```
Comment (njhill) [inline vllm/entrypoints/openai/rpc/server.py:103]: It might be good for this to be `logger.exception(...)` so that the stacktrace is logged.

But don't want to trigger another whole round of CI tests just for this, we can address it in follow-on cleanup of the zmq decoupling that's being done anyhow.

**Comment by njhill** [comment]
Thanks again @joerunde!
